chore(tests) refactor E2E kustomize handling

rainest opened the pull request:
**What this PR does / why we need it**:

Refactor E2E kustomize patching to use structs. This allows contsructing kustomizations with patches or other complex features more easily.

Add a patch to increase the controller init retries.

**Special notes for your reviewer**:

This was an attempt to address E2E failures where Kong fails to ever listen (it simply never makes any `bind()` call) by allowing more retries. It didn't quite work; we need at least extended liveness probes also. Extending the kustomize support and DRYing those functions is useful on its own.

Testing other things to try and reduce timeouts other ways but going ahead and submitting this on its own. https://github.com/Kong/kubernetes-ingress-controller/actions/runs/3147412453/attempts/1 is the "didn't fall over entirely" test.

randmonkey reviewed (commented):
Generally OK. but need to address some linter complaints.

rainest commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,146 @@
+	// annoying single-line strings, which don't break. Alternative is to carve out more of the krusty innards and
+	// skip writing to disk, and somehow load the bytes directly into the kustomize builder.
+
+	increaseControllerInitRetriesPatch = "- op: add\n  path: /spec/template/spec/containers/1/env/-\n  value:\n    name: CONTROLLER_KONG_ADMIN_INIT_RETRIES\n    value: \"%d\""
```
Are you able to get this to work? That's the reason for the comment above: I attempted to do this originally and, for some reason, this results in the string somehow marshaling to a JSON version of the YAML, and the parser rejects it after: https://kubernetes.slack.com/archives/C9A5ALABG/p1664398494955879

There may be some variant with multiline strings that it does accept, but I couldn't find it. The single-line version is annoying, but works.

rainest replied:
Okay, this was an indentation issue. Raw YAML blocks in structs that could support a nested struct just fine were a mistake.

rainest commented:
Added the liveness changes after a test run with them added directly to the manifests produced good results--no Kong failing to start errors (though still plenty of KIND failing to start errors). I did use 30s initial in that run but that's probably excessive, so dropped to 15 for the actual change.

Updated trial in https://github.com/Kong/kubernetes-ingress-controller/actions/runs/3154144363

pmalek commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,168 @@
+
+// patchLivenessProbes patches the given container's liveness probe, replacing the initial delay, period, and failure
+// threshold.
+func patchLivenessProbes(baseManifestReader io.Reader, container, initial, period, failure int) (io.Reader, error) {
```
I know that this might look a lot easier on the eyes like this but I'd much rather see us use `time.Duration` for expressing time duration (which can then extract seconds via: `.Seconds()`).

Especially given the fact that there's a `container` argument which has the same type just next to the duration arguments 😬

pmalek commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,168 @@
+// patchControllerStartTimeout adds or updates the controller container CONTROLLER_KONG_ADMIN_INIT_RETRIES and
+// CONTROLLER_KONG_ADMIN_INIT_RETRY_DELAY environment variables with the provided values, and returns the modified
+// manifest.
+func patchControllerStartTimeout(baseManifestReader io.Reader, tries int, delay string) (io.Reader, error) {
```
... and here on the other hand `delay` is expressed as `string`? Why the difference with `period` using `int` below? Can we also use `time.Duration` for `delay` here?

rainest replied:
My design for both approaches it as if you're editing the actual manifests (that's the process I'm working from), so the types are the types from the original JSON/YAML.

The two are targeting separate pieces of configuration that chose different representation of time durations:
- The Kubernetes livenessProbe type uses seconds only for all its durations, so those got ints.
- The controller `--kong-admin-init-retry-delay` uses a pflag duration, so the strings understood by https://pkg.go.dev/time#ParseDuration

The other option was to make the arg a single dedicated type partially modeling the patched resource, I just didn't bother since I figured we wouldn't look at these often; they're single-purpose wrappers whose values we hopefully don't need to adjust once they make E2E happy.

I kinda prefer the "it's the same value you stick in the manifest" approach but don't really care. Update makes them the native Go types. LMKWYT

pmalek replied:
Someone once said: you write code once but will read it multiple times.

I get what you mean by using the same (as in verbatim) values as in the manifests but OTOH these are not manifests: these are Go functions that can (and should) leverage the type system to ease the reader's mental burden.

So I'd much rather see us use appropriate types instead as suggested before.

pmalek replied:
The reason I'm advocating this change is to make call sites like these:

```
patchLivenessProbes(manifestsReader, 1, 15, 3, 10)
```

more understandable at first glance.

pmalek commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,169 @@
+	if err != nil {
+		return nil, err
+	}
+	kustomization.Bases = []string{"base.yaml"}
```
As indicated earlier: if we can I'd go for using `Resources` here. Even the field itself has a `Deprecation` notice.

Better to address it now then at some point in the future when kustomize API will remove this completely.

pmalek commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,169 @@
+// and returns the modified manifest.
+func patchControllerImage(baseManifestReader io.Reader, image, tag string) (io.Reader, error) {
+	kustomization := types.Kustomization{
+		Bases: []string{"base.yaml"},
```
Ditto

pmalek commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,169 @@
+// and returns the modified manifest.
+func patchKongImage(baseManifestsReader io.Reader, image, tag string) (io.Reader, error) {
+	kustomization := types.Kustomization{
+		Bases: []string{"base.yaml"},
```
Ditto

rainest commented:
https://github.com/Kong/kubernetes-ingress-controller/actions/runs/3178282634 after updating to use the equivalent functionality now in KTF. vet is happy with it so shouldn't be more E2E breakage than usual.

pmalek commented on test/e2e/kustomizations.go:
```diff
@@ -0,0 +1,119 @@
+	return kubectl.GetKustomizedManifest(kustomization, baseManifestReader)
+}
+
+// patchKongImage replaces the kong and kong/kong-gateway images in a manifest with the provide image and tag,
```
```suggestion
// patchKongImage replaces the kong and kong/kong-gateway images in a manifest with the provided image and tag,
```

pmalek commented on test/e2e/utils_test.go:
```diff
@@ -170,96 +167,26 @@ func getTestManifest(t *testing.T, baseManifestPath string) (io.Reader, error) {
-
-func patchKongImage(baseManifestsReader io.Reader, image string, tag string) (io.Reader, error) {
-	workDir, err := os.MkdirTemp("", "kictest.")
+	manifestsReader, err = patchLivenessProbes(manifestsReader, 1, 15, 3, 10)
```
yikes, this one would wait 3 and 10 nanoseconds 😬 

```suggestion
	manifestsReader, err = patchLivenessProbes(manifestsReader, 1, 15, time.Second*3, time.Second*10)
```
